### PR TITLE
Implementation of the method body writer

### PR DESCRIPTION
**Pull request by astherath**
- init work on the file writer mod
- added new methods to extract all endpoint data from toml easily
- finished the header writer mod to abstract the creation of import strings
- small refactoring and changing of the body writer. starting full implementation now
- added common http block trait for all future builder implementations
- made the endpoint cloneable and implemented the base set of trait methods for the BodyBuilder mod
- had to make most of the  (and related) fields public, implemented method to get actix macro string
- added new field to the  struct (and toml) for the method name. Finished the method signature impl.
- first pass at implementing the method body writer. Technically works but uses unlinked  call.
- now header string gen. is no longer based off of enpoint (not needed anyway). also fixed/updated imports list and tidied up some code

The idea of this module is to write all of the boilerplate actix method (i.e. route/endpoint) code correctly.

Altough the original name of the structs and etc. was "...Writer" it was swapped to "...Builder" because the module isnt actually performing any I/O.

### CODE REVIEW

File: sample.toml [hunk 1]
```diff
@@ -15,3 +16,14 @@ name = "users"
...
+
+	[[groups.endpoints]]
+	route = "/find/all"
+	name = "find_all_users"
+	http_verb = "Get"
+	success_code = 200
+	return_model = """\
+	struct UserResponseList {
+		users: Vec<UserResponse>,
+	}"""
```
Comment (astherath) [inline sample.toml:29]: new endpoint to test the code against an actual list

File: src/writers/file_writer/body_writer/mod.rs [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+    /// Returns the actual method signature string for the actix method
+    fn method_signature_string(&self) -> String;
+
+    /// Returns the body string of the actix method.
+    ///
+    /// This method can call many helper methods as long as the result
+    /// returned is final and output-ready.
+    fn method_body_string(&self) -> String;
+}
```
Comment (astherath) [inline src/writers/file_writer/body_writer/mod.rs]: The actual code is pretty straightforward but the structure might be a little weird upon first glance so i'll break it down (although if it isnt readable without this, it's a bad smell anyway)

### the `BodyBuilder` trait
the name is funny by accident, but the idea is very solid.

All of the HTTP verbs will have slightly different implementations of the same string builder methods (macro, method sign., etc.), so adopting a trait with their different phrasing makes the system more akin to a polymorphic handler pattern vs. what would be something closer to an abstract factory (if we were to implement them all one by one and then have the user choose).

### The actual string builder methods
Writing code by formatting strings is hacky at best and a nightmare at worst, so great care went into making this system be as clean, legible, and maintainable as possible.

Basically, every major part of an `actix` endpoint method was broken down into its components and the components are handled as cleanly as possible.

This means that to extend a certain component, only a single function (per http verb struct) is edited; this is fantastic.

This also means that if we would like to add a totally _new_ component, we only need to edit the `trait` block and the compiler will force us to implement the new methods (also great).

File: src/writers/file_writer/header_writer/mod.rs [hunk 3]
```diff
@@ -28,15 +26,25 @@ impl HeaderBuilder {
...
+            "middleware",
+            "web",
+            "App",
+            "HttpRequest",
+            "HttpServer",
+            "http",
+            "get",
+            "HttpResponse",
+            "Responder",
+        ];
```
Comment (astherath) [inline src/writers/file_writer/header_writer/mod.rs:39]: I'm split on how much i like this approach (importing all of the used `actix` structs/methods by name).

On one hand, it makes the code generator strings much easier to write because its just less verbeage.

On the other hand, this is pretty ugly and probably doesn't even make an impact in the end product once compiled.

So far i'll stick with this method but if the import errors get unwieldy then this is the first place that's getting the refactor hammer

File: src/writers/file_writer/mod.rs [hunk 4]
```diff
@@ -1,13 +1,22 @@
...
 
 pub fn write(endpoints: &Vec<&Endpoint>) -> std::io::Result<()> {
     let writer = http_get_writer::HTTPGetEndpointBuilder::new();
+
+    let mut full_output_string = String::new();
+
+    // get and concat the header string to the output string
+    let header_string = header_writer::HeaderBuilder::get_header_string();
+    full_output_string.push_str(&header_string);
+
```
Comment (astherath) [inline src/writers/file_writer/mod.rs]: header string only needs to be written once (duh), previously was getting written per each endpoint
